### DISCUSSION TITLE
About "No_contig"

### DISCUSSION THREAD

**Discussion by TsingYoo** [tuonglab/dandelion#291] [Q&A] [answered]
Dear Professor Kelvin,
    I've used Dandelion for a few days, and discovered that this is a really powerful software. However, there is a question come during using Dandelion. I found that there were a large number of "No_contig" cells  appeared not only when I read your documentation and examples, but also appeared when I processed my own data. But when I checked my data, I found that the match ratio of the VDJ from cellranger outputs is quite high compared with the large number of "No_contig" cells. So could you please tell me the reason of the conflict between "high match ratio of VDJ from cellranger outputs" and "low match ratio of VDJ from Dandelion"?
    Thank you very much!

**Comment by zktuong** [comment, marked as answer]
Hi @TsingYoo, thank you for your interest in Dandelion. Can i ask you what do you mean by `high match ratio of VDJ from cellranger outputs`? Do you mean if you look at the barcode of a cellranger output file (e.g. filtered_contig_annotations.csv), you see that there are more cell barcodes with contigs?

Dandelion only adds a `No_contig` when the cell in the single-cell gene expression data does not match with any cell barcodes found in the VDJ data. This can arise from a number of reasons, but most likely due to incorrect barcode naming between GEX and VDJ data (especially if the cell barcodes are changed by the user).

For example, a common practice at sanger is GEX and VDJ library ids have separate ids, leading to a situation where the outputs have differen folder names, which can be confusing when adding the foldernames to the cell barcodes as a simple prefix.

**Reply by TsingYoo** [reply]
Thank you for your reply! 

Your explanation of "high match ratio of VDJ from cellranger outputs" is exactly what I mean!
After reading your answer, I checked the barcode and found that the barcode is consistent with the Dandelion output file named "filtered_contig_dandelion.tsv", so I think the question might not caused by the unmatched barcode.
What's more, I read your tutorial documentation, and found that there are also a large proportion of cells with "No_contig" label from the sample files, such as the following image "https://sc-dandelion.readthedocs.io/en/latest/_images/notebooks_Q3-analysis_9_0.png". As you can see from the above picture, in the third UMAP image, "No_contig" cells possessed large area of the whole picture, just as what has happened to my own data.
I am very confused about it, is this an inevitable phenomenon?

**Reply by zktuong** [reply]
Hi,

That’s because most of the data aren’t T or B cells e.g. macrophages which don’t have TCR/BCR. Only T/B cells will get tagged with their contigs, otherwise it’s no contigs for all the other cell-types.

**Reply by TsingYoo** [reply]
Professor Kelvin,
Based on your guidance, I successfully solved the former question after processing my data again. Maybe I made some mistakes when running Dandelion from R which caused the question. Thank you again for your patience! 
Besides, I noticed that there is a label named "Extra pair" in "chain_status" and a label named "TRB + Extra VJ" in "locus_status", can I ask you the specific meaning of "Extra pair" and "Extra VJ”?

**Reply by zktuong** [reply]
no worries! 

`Extra pair` means whenever there is a possibility of more than 1 pair of chains e.g. `TRB + first TRA` or `TRB + second TRA`. That's also what the `Extra VJ` in the other column mean - there's more than 1 VJ (TRA) chain

**Reply by TsingYoo** [reply]
Do you mean that "Extra pair" and "Extra VJ" don't have exact significance?
We have some T cell samples from healthy controls and patients, and we found that there are more cells with the label "Extra pair" and "Extra VJ" in patients than in healthy controls. Could it caused by the illness, or just a coincidence?

**Reply by liezeltamon** [reply]
Hi @zktuong,

Thanks again for this tool! So in my data, I have a cluster of cells that cluster with other B-cells but most of its members have no V(D)J contigs (also no TCR contigs) and so I wanted to confirm my understanding of how the `has_contig` column is generated.  So I think if I don't specify `--filter_to_high_confidence` when doing the reannotation, any cell with vdj `umi_count` > 0 will be considered to have contig? i.e. no extra steps looking at ambient signal from empty cells then deciding on a `umi_count` threshold based on that?, which is possible when starting with the cellranger `all_contig_annotations.csv` I think?

**Reply by zktuong** [reply]
Hi @liezeltamon, the `has_contig` assignment simply asks (from a cell's perspective) whether there are any contigs (regardless of UMI) in the airr data that would have the matched corresponding cell barcode. So if you start from `all_contig_annotations.csv`, and without any further manipulation, then yes it should populate accordingly as long as the cell barcode (or `cell_id`) are present.

UMI count filtering is not something dandelion does internallly (other than to break ties when running `check_contigs` for cells with multiple contigs).

**Reply by zktuong** [reply]
also just to add that the `has_contig` check runs in the first couple of lines before any of the other checks as it simply checks for presence/absence of cell barcodes in the airr data and anndata object. It gets updated later once the rest of the QC finishes (if this qc process removes ambiguous contigs)
